### PR TITLE
[crmsh-4.5] Fix: Raise an exception as a rapid return of ssh-related operations to prevent hang  (bsc#1228899)

### PR DESCRIPTION
**Pull request by liangxin1300**
## Problem
In some environments with a network firewall dropping ssh packets, any ssh operation could hang. 

## Solution
- Add the 'core.no_ssh' option in crm.conf. 
  "no" is the default. 
  "yes" means all ssh-related operations are disabled and will directly return on purpose.
- Define a 'utils.NoSSHError' exception and an ssh wrapper function. Raise the exception when `no_ssh` is `yes`.
- Catch `utils.NoSSHError` in `main` so that SSH-related commands will log a general error message: `ERROR: ssh-related operations are disabled. crm report works in local mode.`
- Catch 'utils.NoSSHError' when running certain commands that call ssh, and provide a more user-friendly message.
- `upgradeutil`, which is called regardless of the command being called, will return without calling ssh, when `no_ssh` is `yes`.
- Other commands that are not related to SSH will continue to function normally.

## SSH related commands
- crm cluster stop --all
- crm cluster start --all
- crm report

### CODE REVIEW

**Review by nicholasyang2022 — COMMENTED** [review]
1. If there are `core.hosts` in `crm.conf`, `_guess_user_for_ssh` will not be called. So this patch will not respect `no_ssh` in `crm.conf` if there are `core.hosts`. Please check `no_ssh` in `UserOfHost.user_pair_for_ssh`.
2. `ssh_command()` is not used by other modules. Why is it a public function?

File: crmsh/utils.py [hunk 1]
```diff
@@ -182,7 +182,7 @@ def _get_user_of_host_from_config(host):
 
     @staticmethod
     def _guess_user_for_ssh(host: str) -> typing.Tuple[str, str]:
-        args = ['ssh']
+        args = [ssh_command()]
```
Comment (liangxin1300) [inline crmsh/utils.py:185]: When there is no `core.hosts` in crm.conf, calling ssh here

File: crmsh/utils.py [hunk 2]
```diff
@@ -2946,7 +2946,7 @@ def subprocess_run_auto_ssh_no_input(cmd, remote=None, user=None, **kwargs):
             user = 'root'
         return su_subprocess_run(
             local_user,
-            'ssh {} {}@{} sudo -H -u {} /bin/sh'.format(constants.SSH_OPTION, remote_user, remote, user),
+            '{} {} {}@{} sudo -H -u {} /bin/sh'.format(ssh_command(), constants.SSH_OPTION, remote_user, remote, user),
```
Comment (liangxin1300) [inline crmsh/utils.py:2949]: When there is `core.hosts` in crm.conf, calling ssh here

Comment (nicholasyang2022) [reply crmsh/utils.py:2949]: So you divided a simple problem unnecessary into 2 parts. If you just check `no_ssh` in `user_pair_for_ssh` and raise an exception when it is `yes`, both `_guess_user_for_ssh` and this line will be skipped.

**Comment by liangxin1300** [comment]
```
15sp5-1:~ # ssh -o ConnectTimeout=5 15sp5-2
ssh: connect to host 15sp5-2 port 22: Connection timed out
15sp5-1:~ # cat /etc/crm/crm.conf |grep -E -A1 "^\[core\]"
[core]
no_ssh = yes
15sp5-1:~ # crm status
Status of pacemakerd: 'Pacemaker is running' (last updated 2024-08-09 10:34:16 +08:00)
Cluster Summary:
  * Stack: corosync
  * Current DC: 15sp5-2 (version 2.1.5+20221208.a3f44794f-150500.6.17.1-2.1.5+20221208.a3f44794f) - partition with quorum
  * Last updated: Fri Aug  9 10:34:16 2024
  * Last change:  Wed Aug  7 14:24:39 2024 by hacluster via crmd on 15sp5-1
  * 2 nodes configured
  * 0 resource instances configured

Node List:
  * Online: [ 15sp5-1 15sp5-2 ]

Full List of Resources:
  * No resources

15sp5-1:~ # crm corosync status
Printing ring status.
Local node ID 1
RING ID 0
	id	= 192.168.122.57
	status	= ring 0 active with no faults
15sp5-1:~ # crm configure show
node 1: 15sp5-1
node 2: 15sp5-2
property cib-bootstrap-options: \
	have-watchdog=false \
	dc-version="2.1.5+20221208.a3f44794f-150500.6.17.1-2.1.5+20221208.a3f44794f" \
	cluster-infrastructure=corosync \
	cluster-name=hacluster
rsc_defaults build-resource-defaults: \
	resource-stickiness=1
15sp5-1:~ # crm cluster stop --all
ERROR: ssh-related operations are disabled. crmsh works in local mode.
INFO: Please try 'crm cluster stop' on each node
15sp5-1:~ # crm report -S
15sp5-1: INFO: The report is saved in ./crm_report-Fri-09-Aug-2024.tar.gz
15sp5-1: INFO: Report timespan: 08/08/24 22:34:00 - 08/09/24 10:34:51
15sp5-1: INFO: Thank you for taking time to create this report.
15sp5-1:~ # crm report -n 15sp5-1
15sp5-1: INFO: The report is saved in ./crm_report-Fri-09-Aug-2024.tar.gz
15sp5-1: INFO: Report timespan: 08/08/24 22:34:00 - 08/09/24 10:34:59
15sp5-1: INFO: Thank you for taking time to create this report.
15sp5-1:~ # crm report
15sp5-1: ERROR: ssh-related operations are disabled. crm report works in local mode.
15sp5-1: INFO: The report is saved in ./crm_report-Fri-09-Aug-2024.tar.gz
15sp5-1: INFO: Report timespan: 08/08/24 22:35:00 - 08/09/24 10:35:12
15sp5-1: INFO: Thank you for taking time to create this report.
15sp5-1:~ # crm cluster start --all
INFO: The cluster stack already started on 15sp5-1
ERROR: ssh-related operations are disabled. crmsh works in local mode.
INFO: Please try 'crm cluster start' on each node
15sp5-1:~ # crm cluster stop
INFO: The cluster stack stopped on 15sp5-1
15sp5-1:~ # crm cluster start --all
ERROR: ssh-related operations are disabled. crmsh works in local mode.
INFO: Please try 'crm cluster start' on each node
```

File: crmsh/main.py [hunk 3]
```diff
@@ -367,6 +367,9 @@ def run():
         else:
             upgradeutil.upgrade_if_needed()
             return main_input_loop(context, user_args)
+    except utils.NoSSHError as msg:
+        logger.error(msg)
```
Comment (nicholasyang2022) [inline crmsh/main.py:371]:
```suggestion
        logger.error('%s', msg)
```

**Review by nicholasyang2022 — APPROVED** [review]

**Pull request merged**